File: android/app/src/main/kotlin/com/woosmap/woosmap_geofencing_rest_sample/GeofencingEventsReceiver.kt
package com.woosmap.woosmap_geofencing_rest_sample

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.util.Log
import com.webgeoservices.woosmapgeofencingcore.database.POI
import com.webgeoservices.woosmapgeofencingcore.database.WoosmapDb
import org.json.JSONObject
import java.io.OutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * Listens for the `com.woosmap.action.GEOFENCE_TRIGGERED` broadcast emitted by
 * the Woosmap Geofencing SDK and forwards the region event to the customer's
 * back-office via a REST call (replacing the Batch SDK call from the Woosmap
 * Batch connector sample).
 *
 * The `regionLog` extra is a JSON string, e.g.:
 * {
 *   "longitude": -0.1337,
 *   "latitude": 51.50998,
 *   "date": 1700824501480,
 *   "didenter": true,
 *   "identifier": "custom-region1",
 *   "radius": 100,
 *   "frompositiondetection": false,
 *   "eventname": "woos_geofence_exited_event",
 *   "spenttime": 75
 * }
 */
class GeofencingEventsReceiver : BroadcastReceiver() {

    companion object {
        private const val TAG = "GeofencingReceiver"

        // TODO: point these at the customer's back-office.
        private const val BACK_OFFICE_URL =
            "https://api.customer-backoffice.example.com/geofence-events"
        private const val API_KEY = "CUSTOMER_API_KEY"
    }

    private val executorService: ExecutorService = Executors.newSingleThreadExecutor()

    override fun onReceive(context: Context?, intent: Intent?) {
        Log.d(TAG, "Received geofence broadcast")
        if (context == null || intent == null) return

        executorService.execute {
            try {
                val regionLog = intent.getStringExtra("regionLog") ?: return@execute
                val regionData = JSONObject(regionLog)

                // Build the base payload from the region log.
                val payload = JSONObject().apply {
                    put("date", regionData.optLong("date"))
                    put("eventName", regionData.optString("eventname"))
                    put("id", regionData.optString("identifier"))
                    put("latitude", regionData.optDouble("latitude"))
                    put("longitude", regionData.optDouble("longitude"))
                    put("radius", regionData.optDouble("radius"))
                    put("didEnter", regionData.optBoolean("didenter"))
                    put("spentTime", regionData.optLong("spenttime"))
                }

                // Enrich with POI attributes if the region maps to a known POI.
                // poi is null for custom regions.
                val poi: POI? = WoosmapDb.getInstance(context)
                    .poIsDAO
                    .getPOIbyStoreId(regionData.optString("identifier"))
                if (poi != null) {
                    payload.put("idStore", poi.idStore)
                    payload.put("name", poi.name)
                    payload.put("city", poi.city)
                    payload.put("zipCode", poi.zipCode)
                    payload.put("distance", poi.distance)
                    payload.put("countryCode", poi.countryCode)
                    payload.put("address", poi.address)
                    payload.put("tags", poi.tags)
                    payload.put("types", poi.types)
                }

                postEvent(payload)
            } catch (ex: Exception) {
                Log.e(TAG, "Failed to handle geofence event: $ex")
            }
        }
    }

    private fun postEvent(payload: JSONObject) {
        var connection: HttpURLConnection? = null
        try {
            connection = (URL(BACK_OFFICE_URL).openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                connectTimeout = 30_000
                readTimeout = 30_000
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
                setRequestProperty("Authorization", "Bearer $API_KEY")
            }

            connection.outputStream.use { os: OutputStream ->
                os.write(payload.toString().toByteArray(Charsets.UTF_8))
            }

            val code = connection.responseCode
            Log.d(TAG, "Back-office responded $code")
        } catch (ex: Exception) {
            Log.e(TAG, "Back-office POST error: $ex")
        } finally {
            connection?.disconnect()
        }
    }
}

File: ios/Runner/GeofencingEventsReceiver.swift
import Foundation
import UIKit
import WoosmapGeofencing

extension Notification.Name {
  static let updateRegions = Notification.Name("updateRegions")
  static let didEventPOIRegion = Notification.Name("didEventPOIRegion")
}

/// Listens for Woosmap geofence region events on iOS and forwards them to the
/// customer's back-office via a REST call (replacing the Batch SDK call from
/// the Woosmap Batch connector sample).
@objc(GeofencingEventsReceiver)
class GeofencingEventsReceiver: NSObject {

  // TODO: point these at the customer's back-office.
  private let backOfficeURL = URL(string: "https://api.customer-backoffice.example.com/geofence-events")!
  private let apiKey = "CUSTOMER_API_KEY"

  @objc public func startReceivingEvent() {
    NotificationCenter.default.addObserver(
      self,
      selector: #selector(POIRegionReceivedNotification),
      name: .didEventPOIRegion,
      object: nil)
  }

  @objc func POIRegionReceivedNotification(notification: Notification) {
    guard let region = notification.userInfo?["Region"] as? Region else { return }

    // Event name convention matches the Woosmap connectors.
    let eventName = region.didEnter
      ? "woos_geofence_entered_event"
      : "woos_geofence_exited_event"

    // Base payload — always present.
    var payload: [String: Any] = [
      "date": ISO8601DateFormatter().string(from: region.date ?? Date()),
      "eventName": eventName,
      "id": region.identifier ?? "",
      "latitude": region.latitude,
      "longitude": region.longitude,
      "radius": region.radius,
      "didEnter": region.didEnter,
      "origin": region.origin ?? ""
    ]

    // Enrich with POI attributes when the region originates from a POI.
    if region.origin == "POI",
       let poi = POIs.getPOIbyIdStore(idstore: region.identifier ?? "") as POI? {
      payload["idStore"]     = poi.idstore ?? ""
      payload["name"]        = poi.name ?? ""
      payload["city"]        = poi.city ?? ""
      payload["zipCode"]     = poi.zipCode ?? ""
      payload["distance"]    = poi.distance
      payload["countryCode"] = poi.countryCode ?? ""
      payload["address"]     = poi.address ?? ""
      payload["tags"]        = poi.tags ?? ""
      payload["types"]       = poi.types ?? ""
    }

    postEvent(payload)
  }

  private func postEvent(_ payload: [String: Any]) {
    // Keep the app alive long enough to finish the request when the event is
    // triggered while the app is in the background.
    var bgTask: UIBackgroundTaskIdentifier = .invalid
    bgTask = UIApplication.shared.beginBackgroundTask(withName: "WoosGeofenceEvent") {
      UIApplication.shared.endBackgroundTask(bgTask)
      bgTask = .invalid
    }

    var request = URLRequest(url: backOfficeURL)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
    request.timeoutInterval = 30

    do {
      request.httpBody = try JSONSerialization.data(withJSONObject: payload, options: [])
    } catch {
      NSLog("Woos: failed to serialize payload: \(error)")
      UIApplication.shared.endBackgroundTask(bgTask)
      return
    }

    let task = URLSession.shared.dataTask(with: request) { _, response, error in
      if let error = error {
        NSLog("Woos: back-office POST error: \(error.localizedDescription)")
      } else if let http = response as? HTTPURLResponse {
        NSLog("Woos: back-office responded \(http.statusCode)")
      }
      if bgTask != .invalid {
        UIApplication.shared.endBackgroundTask(bgTask)
        bgTask = .invalid
      }
    }
    task.resume()
  }

  @objc public func stopReceivingEvent() {
    NotificationCenter.default.removeObserver(self, name: .didEventPOIRegion, object: nil)
  }
}
